hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/WALEntryStream.java

```diff
@@ -255,15 +255,21 @@ private void dequeueCurrentLog() throws IOException {
    * Returns whether the file is opened for writing.
    */
   private boolean readNextEntryAndRecordReaderPosition() throws IOException {
+    long prePos = reader.getPosition();
```

Comment:
I think this is a nice improvement that we can return earlier when there is no data can be read. But I do not get the point that why here if we try to read more data can lead to a problem? We could still hit EOFException when the remaining data is not enough for a whole entry?

Reply:
Thanks for reviewing.
I haven't found out yet that we read an incomplete entry. EOFException is thrown by `org.apache.hadoop.fs.FSDataInputStream#seek`.
After reading the uncommitted data, we will reopen `FSDataInputStream` and seek back. If the position where seek back is the sync length, it may cause EOF. As the 3 copies of the hdfs block being written are inconsistent, even if we can read the data from a specific position in one copy, it may not be read in the other one.
It's not a big problem if it just throws EOF as we'll retry. The big problem is that if we have read some entries into the `WALEntryBatch` and increased the [`totalBufferUsed`](https://github.com/apache/hbase/blob/308cd729d23329e6d8d4b9c17a645180374b5962/hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReader.java#L78), and the `totalBufferUsed` is not subtracted after throwing EOF, all peers will eventually block completely.

Reply:
> It's not a big problem if it just throws EOF as we'll retry. The big problem is that if we have read some entries into the WALEntryBatch and increased the totalBufferUsed, and the totalBufferUsed is not subtracted after throwing EOF, all peers will eventually block completely.

One of our customers seems to be consistently reaching this EOF problem, per below exception trace. 

`WARN org.apache.hadoop.hbase.regionserver.wal.ProtobufLogReader: Encountered a malformed edit, seeking back to last good position in file, from 31912404 to 31912265
java.io.EOFException: EOF while reading 106 WAL KVs; started reading at 31912338 and read up to 31912404
at org.apache.hadoop.hbase.regionserver.wal.ProtobufLogReader.readNext(ProtobufLogReader.java:397)
at org.apache.hadoop.hbase.regionserver.wal.ReaderBase.next(ReaderBase.java:98)
at org.apache.hadoop.hbase.regionserver.wal.ReaderBase.next(ReaderBase.java:86)
at org.apache.hadoop.hbase.replication.regionserver.WALEntryStream.readNextEntryAndRecordReaderPosition(WALEntryStream.java:262)
at org.apache.hadoop.hbase.replication.regionserver.WALEntryStream.tryAdvanceEntry(WALEntryStream.java:176)
at org.apache.hadoop.hbase.replication.regionserver.WALEntryStream.hasNext(WALEntryStream.java:101)`

Reply:
Checking the code, we have a special logic in ReplicationSourceWALReader for dealing with EOFException. Please see ReplicationSourceWALReader.handleEofException.

Reply:
Thanks @wchevreuil for reviewing and sharing the case. This looks like it reads an incomplete entry?
> Then I do not think the fix here is enough?

As @Apache9 said, this fix may not enough, it may not cover that reading incomplete entry caused by like RS down. I'm digging deeper into the root cause.

Reply:
@wchevreuil for the above stacktrace you mentioned, which version the cluster is running on?

Reply:
FYI @apurtell

Reply:
> One of our customers seems to be consistently reaching this EOF problem, per below exception trace.
> 
> `WARN org.apache.hadoop.hbase.regionserver.wal.ProtobufLogReader: Encountered a malformed edit, seeking back to last good position in file, from 31912404 to 31912265 java.io.EOFException: EOF while reading 106 WAL KVs; started reading at 31912338 and read up to 31912404 at org.apache.hadoop.hbase.regionserver.wal.ProtobufLogReader.readNext(ProtobufLogReader.java:397) at org.apache.hadoop.hbase.regionserver.wal.ReaderBase.next(ReaderBase.java:98) at org.apache.hadoop.hbase.regionserver.wal.ReaderBase.next(ReaderBase.java:86) at org.apache.hadoop.hbase.replication.regionserver.WALEntryStream.readNextEntryAndRecordReaderPosition(WALEntryStream.java:262) at org.apache.hadoop.hbase.replication.regionserver.WALEntryStream.tryAdvanceEntry(WALEntryStream.java:176) at org.apache.hadoop.hbase.replication.regionserver.WALEntryStream.hasNext(WALEntryStream.java:101)`

@wchevreuil, this problem seems not throw EOFException in `WALEntryStream.readNextEntryAndRecordReaderPosition`, it may be another problem?